x-pack/legacy/plugins/apm/public/components/app/TraceLink/index.tsx

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { EuiEmptyPrompt } from '@elastic/eui';
+import moment from 'moment-timezone';
+import styled from 'styled-components';
+import { Redirect } from 'react-router-dom';
+import { useFetcher } from '../../../hooks/useFetcher';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { convertTo, getDateDifference } from '../../../utils/formatters';
+import { Transaction } from '../../../../typings/es_schemas/ui/Transaction';
+import { generatePath } from '../Main/route_config/utils';
+import { RouteName } from '../Main/route_config/route_names';
+
+interface RedirectType {
+  pathname: string;
+  search: string;
+}
+
+const CentralizedContainer = styled.div`
+  height: 100%;
+  display: flex;
+`;
+
+const redirectToTransactionDetailPage = (
+  traceId: string,
+  transaction: Transaction
+): RedirectType => {
+  const pathname = generatePath(RouteName.TRANSACTION_NAME, {
+    serviceName: transaction.service.name
+  });
+
+  const {
+    id: transactionId,
+    name: transactionName,
+    type: transactionType
+  } = transaction.transaction;
+
+  let search = `?traceId=${traceId}&transactionId=${transactionId}&transactionName=${transactionName}&transactionType=${transactionType}`;
+
+  const diff = getDiffDays(transaction.timestamp.us);
+  if (diff > 0) {
+    search = `${search}&rangeFrom=now-${diff}d&rangeTo=now`;
+  }
+  return { pathname, search };
+};
+
+const getDiffDays = (microseconds: number) => {
+  const start = moment(
+    convertTo({
+      unit: 'milliseconds',
+      microseconds
+    }).convertedValue
+  );
+
+  return getDateDifference(start, moment(), 'days');
+};
+
+const redirectToTracePage = (traceId: string): RedirectType => ({
+  pathname: generatePath(RouteName.TRACES),
+  search: `?kuery=${encodeURIComponent(`trace.id : "${traceId}"`)}`
+});
+
+export const TraceLink = () => {
+  const { urlParams } = useUrlParams();
+  const { traceIdLink: traceId } = urlParams;
+  const { data } = useFetcher(
+    callApmApi => {
+      if (traceId) {
+        return callApmApi({
+          pathname: '/api/apm/transaction/{traceId}',
+          params: {
+            path: {
+              traceId
+            }
+          }
+        });
+      }
+    },
+    [traceId]
+  );
+
+  if (traceId && data) {
+    let to: RedirectType;
+    if (data.total.value === 0) {
+      to = redirectToTracePage(traceId);
+    } else {
+      to = redirectToTransactionDetailPage(traceId, data.hits[0]._source);
+    }
+
+    return <Redirect to={to} />;
+  }
+
+  return (
+    <CentralizedContainer>
+      <EuiEmptyPrompt iconType="apmTrace" title={<h2>Fetching trace...</h2>} />
+    </CentralizedContainer>
+  );
+};
```

Comment:
I think we should drop `generatePath` for now. It's possibly something we want to do later but it's too big an effort to bring into this PR.

I also think we should drop the automatic from/to ranges, and instead have the consumer pass them in:
```
/link-to/trace/1337?start=...&end...
```

I think something like this will be better (untested):
```ts
// use `url.format` and `fromQuery` instead of manually constructing the url
function getTransactionDetailsPageUrl(transaction: Transaction) {
  return url.format({
    pathname: core.http.basePath.prepend('/app/apm'),
    search: fromQuery({
      traceId: transaction.trace.id,
      transactionId: transaction.id,
      transactionName: transaction.name,
      transactionType: transaction.type
    })
  });
}

export const TraceLink = () => {
  const { urlParams } = useUrlParams();
  const { traceIdLink: traceId } = urlParams;
  const { data = { transaction: null }, status } = useFetcher(
    callApmApi => {
      if (traceId) {
        return callApmApi({
          pathname: '/api/apm/transaction/{traceId}',
          params: {
            path: {
              traceId
            }
          }
        });
      }
    },
    [traceId]
  );

  // use `status` to determine whether the transaction was loaded
  if (status === 'success') {
    const to = data.transaction
      ? getTransactionDetailsPageUrl(data.transaction)
      : getTracePageUrl(traceId);
    return <Redirect to={to} />;
  }

  return (
    <CentralizedContainer>
      <EuiEmptyPrompt iconType="apmTrace" title={<h2>Fetching trace...</h2>} />
    </CentralizedContainer>
  );
};
```

Reply:
I created the generatePath to avoid hardcoding the URL, as the pathname of transaction details should be `/services/:serviceName/transactions/view`.
But that's fine if you want to get rid of it, then instead I'll do:
```
return url.format({
    pathname: `/services/${serviceName}/transactions/view`,
    search: fromQuery({
      traceId: transaction.trace.id,
      transactionId: transaction.id,
      transactionName: transaction.name,
      transactionType: transaction.type
    })
  });```

Reply:
Yeah, as I said, I really like that idea and I see the advantage of it but I think it's a slightly bigger task that we should tackle separately

Reply:
@sqren should the `start` and `end` query params, filter the transaction index on ES? or only applied on the Transaction Details Page?

Reply:
I think so.
I haven't thought too much about it tbh. My intention was to go for the simplest approach first. I think it would make sense to involve @jasonrhodes to align on the link api:

```
/link-to/trace/1337?start=...&end...
```

(btw I don't know if it should be `start/end`, `to/from` or `rangeFrom/rangeTo`. We should use whatever is more consistent.

Reply:
Sounds like you are trying to decide on what the path and query params should be for linking directly to a trace, right? And that having some kind of function to generate that link is out of scope for this PR?

If so, I'd say the query param names and path you choose doesn't matter much for now, and that when you tackle the "generatePath" function, you can move it to the observability plugin where it will also have access to our core hooks. In there, it can generate the right path but also provide types for the incoming arguments, so whoever uses `generateApmTraceLink()` or whatever the function is called will have TS telling them what args they should pass, and then the string link will be returned. (TBD if we also want the "TraceLink" component to live centrally for shareable convenience).

Does that make sense? Am I responding to the right thing here? :D

Reply:
> Am I responding to the right thing here? :D

Haha, no! ;) I should have been more clear. 
As a consumer of the link function we are building: when you link from logs app to the apm app, and you want to link to a particular trace, in addition to supplying the `traceId` would it make sense for you to supply the time range as well?

**Option 1**
```js
getTraceLink(traceId, start, end) // returns: `/app/apm#/link-to/trace/1337?start=123&end=456`
```

or

**Option 2**
```js
getTraceLink(traceId) // returns: `/app/apm#/link-to/trace/1337`
```

In the latter case we would have to come up with a time range. We could do that by padding the `transaction.@timestamp` with +/- 5 minutes. That would work but it feels a little too magic. Normally when navigating between apps Kibana should keep the time range - so I'd prefer the consumer specified it. WDYT?